### PR TITLE
Refactor LightningDistributedDataParallel

### LINKED ISSUE #4630: Keeping DDP override in sync with upstream torch
From @ananthsub:
how should Lightning keep its [DDP override](https://github.com/PyTorchLightning/pytorch-lightning/blob/3d202f9ecc4137b08cb5b1ac15af276456fcfaaf/pytorch_lightning/overrides/data_parallel.py#L154-L210) in sync with the upstream torch [DistributedDataParallel](https://github.com/pytorch/pytorch/blob/d478605dec65a746d41506b23693d6013bfa11b2/torch/nn/parallel/distributed.py#L646-L697)? these implementations have now diverged. I think this leads to performance degradations with Lightning + gradient accumulations, since the require_backward_grad_sync attribute isn't checked before the backwards pass

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Fixes #4630
cc @ananthsub 

The motivation behind this refactor of the DDP wrapper is that we get all (future) improvements from upstream pytorch DDP, and the user can comfortably subclass the pytorch wrapper if they want to.

TODO: 
- [x] Docs
- [x] add unit tests for new class to keep coverage up
- [x] add deprecation for old class?

### CODE REVIEW

File: pytorch_lightning/overrides/data_parallel.py [hunk 1]
```diff
@@ -151,74 +152,98 @@ def parallel_apply(self, replicas, inputs, kwargs):
...
+#                 self.reducer.prepare_for_backward([])
+#
+#     def reducer_reset_hooks(self):
+#         self._reducer_prepared_for_backwards = False
 
-        if output is None:
-            warn_missing_output(f'{fx_called} returned None. Did you forget to return an output')
-        return output
 
-    def reducer_prepare_for_backwards(self, output):
```
Comment (awaelchli) [inline pytorch_lightning/overrides/data_parallel.py:207]: I'm not sure where this should go, it requires the reducer from ddp

Comment (awaelchli) [reply pytorch_lightning/overrides/data_parallel.py:207]: maybe a hook?
https://pytorch.org/docs/stable/generated/torch.nn.Flatten.html?highlight=hook#torch.nn.Flatten.register_forward_pre_hook

**Comment by ananthsub** [comment]
fyi @pritamdamania87

File: pytorch_lightning/overrides/data_parallel.py (hunk 1 above)
Comment (awaelchli) [reply pytorch_lightning/overrides/data_parallel.py:207]: @SeanNaren do you remember the conversation we started in #4976 about this? You had an idea there, I'm trying to understand it, maybe you can explain again? :)

File: pytorch_lightning/plugins/ddp_plugin.py [hunk 2]
```diff
@@ -136,23 +137,25 @@ def get_model_from_plugin(
...
 
     @contextmanager
-    def block_backward_sync(self, model: LightningDistributedDataParallel):
+    def block_backward_sync(self, model: DistributedDataParallel):
         """
         Blocks ddp sync gradients behaviour on backwards pass.
         This is useful for skipping sync when accumulating gradients, reducing communication overhead
         Returns: context manager with sync behaviour off
         """
         yield model.no_sync()
```
Comment (ananthsub) [inline pytorch_lightning/plugins/ddp_plugin.py:153]: do we need `block_backward_sync` still? can we directly call `model.no_sync()`

File: pytorch_lightning/overrides/data_parallel.py [hunk 3]
```diff
@@ -151,74 +152,98 @@ def parallel_apply(self, replicas, inputs, kwargs):
...
 
-class LightningDistributedDataParallel(DistributedDataParallel):
-    """
-    Override the forward call in lightning so it goes to training and validation step respectively
-    """
-    PREPARE_FOR_BACKWARDS = True
-
-    def parallel_apply(self, replicas, inputs, kwargs):
-        return parallel_apply(replicas, inputs, kwargs, self.device_ids[:len(replicas)])
+class LightningDistributedWrapper(torch.nn.Module):
```
Comment (ananthsub) [inline pytorch_lightning/overrides/data_parallel.py:155]: for later: we should add comments for how this class is to be used

File: pytorch_lightning/plugins/ddp_plugin.py [hunk 4]
```diff
@@ -63,8 +64,8 @@ def configure_ddp(self, model, device_ids):
         self._ddp_kwargs["find_unused_parameters"] = self._ddp_kwargs.get(
             "find_unused_parameters", True
         )
-        model = LightningDistributedDataParallel(
-            model,
+        model = DistributedDataParallel(
+            module=LightningDistributedWrapper(model),
```
Comment (ananthsub) [inline pytorch_lightning/plugins/ddp_plugin.py:68]: the docstring above needs to be updated for the return type

File: pytorch_lightning/plugins/ddp_plugin.py [hunk 5]
```diff
@@ -63,8 +64,8 @@ def configure_ddp(self, model, device_ids):
         self._ddp_kwargs["find_unused_parameters"] = self._ddp_kwargs.get(
             "find_unused_parameters", True
```
Comment (ananthsub) [inline pytorch_lightning/plugins/ddp_plugin.py:65]: i'm not sure why this is the default. this incurs a perf hit and is different from the DDP default

**Review by ananthsub — COMMENTED** [review]
exciting progress! are there other spots in lightning where `LightningDistributedDataParallel` is referenced? i imagine this will also provide a perf benefit as we can take advantage of improvements in DDP's `forward` automatically now.

- Does the same pattern apply for data parallel?
- (down the line) what if someone wants to wrap only some modules in their lightning module with DDP? what further changes would that require here?

**Comment by awaelchli** [comment]
> are there other spots in lightning where LightningDistributedDataParallel is referenced?

in the end I will do a search across the codebase, but I think it's more or less isolated to the plugins and accelerators and in most places in Lightning we operate on model.module.

> Does the same pattern apply for data parallel?

yes I believe so. currently the biggest difference I see is the custom gather function for Results objects, but since we dropped support we should also be able to get rid of it and just use the native torch implementation for DP.

> (down the line) what if someone wants to wrap only some modules in their lightning module with DDP? what further changes would that require here?

oh, that sounds adventurous! Not sure if in this scenario the wrapper in this PR would stand in the way, because of the order of nesting it implies. Need to think about it more.


At the moment the PR is blocked by the problem of manual optimization. It seems the custom code for calling the reducer is necessary for Lightning's manual backward. I don't fully understand it, I hope @SeanNaren can give me some hints.

File: pytorch_lightning/plugins/ddp_plugin.py (hunk 2 above)
Comment (awaelchli) [reply pytorch_lightning/plugins/ddp_plugin.py:153]: the only reason I can think of why it is there is so the user can override this method in their own plugin, though there is not much customization they can do to this context manager :)

File: pytorch_lightning/plugins/ddp_plugin.py (hunk 5 above)
Comment (awaelchli) [reply pytorch_lightning/plugins/ddp_plugin.py:65]: It was added in this PR: #4382, there must be a good reason, @SeanNaren?

**Comment by SeanNaren** [comment]
Nice! I like the overall flow of this. Just to make it clear for me, all we're doing is adding a higher level torch module on top of the DDP/DP modules? I have to ask; was there any reason this wasn't done initially? I may have missed something in this case.

Regarding manual optimization, the additional logic is due to the backward hooks being added after forward is finished. Right now we assume that if you override `training_step` in manual optimization, you can do everything you like within the function. Before the fix, this meant that if you called `.backward()` in training_step, gradients were no synced as no autograd hooks were added.

We added the additional logic as a temp fix for this, but I really don't like the additional logic. I haven't had time to figure out a cleaner solution than what we added, but our current solution boils down to calling `reducer.prepare_for_backwards(...)` manually which adds the hooks (optionally using the loss to add hooks to modules that actually require it iirc).

File: pytorch_lightning/plugins/ddp_plugin.py (hunk 5 above)
Comment (tchaton) [reply pytorch_lightning/plugins/ddp_plugin.py:65]: Hey @awaelchli. Not sure Will looked into the default for `find_unused_parameters`. Let's stick to Pytorch default which is False right ?

Comment (SeanNaren) [reply pytorch_lightning/plugins/ddp_plugin.py:65]: It should be false, it is only recommended to do this if necessary: https://pytorch.org/docs/stable/notes/ddp.html#internal-design

**Comment by tchaton** [comment]
> Nice! I like the overall flow of this. Just to make it clear for me, all we're doing is adding a higher level torch module on top of the DDP/DP modules? I have to ask; was there any reason this wasn't done initially? I may have missed something in this case.
> 
> Regarding manual optimization, the additional logic is due to the backward all reduce hooks being added **after** forward is finished. Right now we assume that if you override `training_step` in manual optimization, you can do everything you like within the function. Before the fix, this meant that if you called `.backward()` in training_step, gradients were not synced as no autograd reduce hooks were added.
> 
> We added the additional logic as a temp fix for this, but I really don't like the additional logic. I haven't had time to figure out a cleaner solution than what we added, but our current solution boils down to calling `reducer.prepare_for_backwards(...)` manually which adds the hooks (optionally using the loss to add hooks to parameters that actually require it iirc).


By setting model.require_backward_grad_sync = False when model.manual_optimization=True and PipeRPCPlugin is used. It might work. Just to  
In Accelerator.backward 

```
if not automatic_optimization and self.ddp_plugin is not None:
    if should_accumulate:
            self.require_backward_grad_sync = False
    else:
        if self.find_unused_parameters:
            self.reducer.prepare_for_backward(list(_find_tensors(closure_loss)))
        else:
            self.reducer.prepare_for_backward([])
```

**Comment by Borda** [comment]
@awaelchli is this ready to land/review? 🐰

**Comment by tchaton** [comment]
Hey @awaelchli,

Resolved on this checkout from this branch: https://github.com/PyTorchLightning/pytorch-lightning/pull/5415

Best,
T.C

**Comment by awaelchli** [comment]
@tchaton amazing, thanks!

Any preferences/suggestions for a better name? 

- LightningDistributedWrapper (current)
- LightningDistributedModule?

File: pytorch_lightning/overrides/data_parallel.py (hunk 3 above)
Comment (awaelchli) [reply pytorch_lightning/overrides/data_parallel.py:153]: done, with example.

File: pytorch_lightning/plugins/ddp_plugin.py (hunk 4 above)
Comment (awaelchli) [reply pytorch_lightning/plugins/ddp_plugin.py:80]: done

File: pytorch_lightning/plugins/ddp_plugin.py (hunk 5 above)
Comment (awaelchli) [reply pytorch_lightning/plugins/ddp_plugin.py]: ok, made it default False to be in line with pytorch DDP.

File: pytorch_lightning/overrides/data_parallel.py [hunk 6]
```diff
@@ -149,74 +150,65 @@ def parallel_apply(self, replicas, inputs, kwargs):
...
-class LightningDistributedDataParallel(DistributedDataParallel):
-    """
-    Override the forward call in lightning so it goes to training and validation step respectively
-    """
-    PREPARE_FOR_BACKWARDS = True
+class LightningDistributedWrapper(torch.nn.Module):
 
-    def parallel_apply(self, replicas, inputs, kwargs):
-        return parallel_apply(replicas, inputs, kwargs, self.device_ids[:len(replicas)])
+    def __init__(self, lightning_module: LightningModule):
```
Comment (Borda) [inline pytorch_lightning/overrides/data_parallel.py:155]:
```suggestion
    def __init__(self, pl_module: LightningModule):
```
consistently with other calls?

File: pytorch_lightning/overrides/data_parallel.py [hunk 7]
```diff
@@ -149,74 +150,65 @@ def parallel_apply(self, replicas, inputs, kwargs):
         return parallel_apply(replicas, inputs, kwargs, self.device_ids[:len(replicas)])
 
 
-class LightningDistributedDataParallel(DistributedDataParallel):
-    """
-    Override the forward call in lightning so it goes to training and validation step respectively
-    """
-    PREPARE_FOR_BACKWARDS = True
+class LightningDistributedWrapper(torch.nn.Module):
```
Comment (Borda) [inline pytorch_lightning/overrides/data_parallel.py:153]: btw, this smells like api change ;]

Comment (awaelchli) [reply pytorch_lightning/overrides/data_parallel.py:153]: ok, deprecation warning and remove in 1.4?

Comment (awaelchli) [reply pytorch_lightning/overrides/data_parallel.py:165]: @Borda I also added a deprecation test, but it is not so simple because it needs torch.distributed to be initialized, so it will add about 3-4 seconds only for that test. Don't know how to do it simpler

File: pytorch_lightning/overrides/data_parallel.py (hunk 1 above)
Comment (tchaton) [reply pytorch_lightning/overrides/data_parallel.py:205]: I have been thinking about this one maybe too much ahah, but I didn't find a better way to do it as `backward` and `optimizer.step` are being called in `training_step` and `DDP reducer` is being called on training_step output

**Review by tchaton — APPROVED** [review]

**Review by ananthsub — APPROVED** [review]
very nice @awaelchli !

File: CHANGELOG.md [hunk 8]
```diff
@@ -47,6 +47,9 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Changed `automatic casting` for LoggerConnector `metrics` ([#5218](https://github.com/PyTorchLightning/pytorch-lightning/pull/5218))
 
 
+- Changed the default of `find_unused_parameters` to `False` in DDP ([#5185](https://github.com/PyTorchLightning/pytorch-lightning/pull/5185))
+
+
```
Comment (ananthsub) [inline CHANGELOG.md:52]: i wonder if you'd like to pull this out to another PR that's super easy to review and merge (just changing the defaults)

Comment (awaelchli) [reply CHANGELOG.md]: sounds good! I reverted the changes and moved them to #5435

File: pytorch_lightning/plugins/ddp_plugin.py (hunk 5 above)
Comment (ananthsub) [reply pytorch_lightning/plugins/ddp_plugin.py:81]: this could be a nice speedup for distributed training jobs. @SeanNaren n00b question: is there a way to verify this on the lightning benchmarks?

**Comment by caleb603** [comment]
https://github.com/caleb603/cuddly-octo-eureka/blob/52a9e1b72ddfeebf7119b277bd25e5e887c4f1d0/.github/workflows/blank.yml#L30

File: pytorch_lightning/overrides/data_parallel.py [hunk 9]
```diff
@@ -150,73 +152,75 @@ def parallel_apply(self, replicas, inputs, kwargs):
...
-                self.reducer.prepare_for_backward(list(_find_tensors(output)))
-            else:
-                self.reducer.prepare_for_backward([])
-
-    def reducer_reset_hooks(self):
-        self._reducer_prepared_for_backwards = False
+
+# In manual_optimization, we need to call reducer prepare_for_backward.
+# TODO: Keep track of Pytorch DDP and update if there is a change
+# https://github.com/pytorch/pytorch/blob/e6779d4357ae94cc9f9fedb83a87eb6126016769/torch/nn/parallel/distributed.py#L692
```
Comment (carmocca) [inline pytorch_lightning/overrides/data_parallel.py:204]:
```suggestion
# Note: Keep track of Pytorch DDP and update if there is a change
# https://github.com/pytorch/pytorch/blob/v1.7.1/torch/nn/parallel/distributed.py#L626-L638
```

File: pytorch_lightning/plugins/ddp_plugin.py [hunk 10]
```diff
@@ -16,27 +16,27 @@
...
-    which in turn forwards all args to `DistributedDataParallel`.
+    This plugin forwards all constructor arguments to :class:`~torch.nn.parallel.DistributedDataParallel`.
 
     Example::
 
         class MyDDP(DDPPlugin):
 
             def configure_ddp(self, model, device_ids):
-                model = MyDDPWrapper(model, device_ids)
+                model = MyDDPWrapper(LightningDistributedModule(model), device_ids)
```
Comment (carmocca) [inline pytorch_lightning/plugins/ddp_plugin.py:39]: question: what is MyDDPWrapper supposed to be here? wouldn't the user do now:

```python
class MyDDPModule(LightningDistributedModule):
    ...
```

and then use it in configure_ddp?

```python
            def configure_ddp(self, model, device_ids):
                model = MyDDPModule(model, device_ids)
```

File: pytorch_lightning/overrides/data_parallel.py (hunk 1 above)
Comment (carmocca) [reply pytorch_lightning/overrides/data_parallel.py:205]: could open a PR in PyTorch to at least move it to a function we can use

File: pytorch_lightning/plugins/ddp_plugin.py (hunk 10 above)
Comment (awaelchli) [reply pytorch_lightning/plugins/ddp_plugin.py:39]: MyDDPWrapper can be a user's subclass of torch.nn.parallel.DistributedDataParallel. 

```python
class MyDDPWrapper(torch.nn.parallel.DistributedDataParallel):
    ...
```
    
and then the usage in configure_ddp is: 

```python
def configure_ddp(self, model, device_ids):
                model = MyDDPWrapper(LightningDistributedModule(model), device_ids)
```

The only requirement Lightning has is that the user wraps their model first with our LightningDistributedModule. That's all. I show the concrete example below in configure_ddp.

Comment (awaelchli) [reply pytorch_lightning/plugins/ddp_plugin.py:39]: By default MyDDPWrapper == torch.nn.parallel.DistributedDataParallel

File: pytorch_lightning/plugins/ddp_plugin.py [hunk 11]
```diff
@@ -150,24 +154,23 @@ def get_model_from_plugin(
         Returns: Reference :class:`LightningModule` within parallel wrapper.
 
         """
-        if isinstance(model, LightningDistributedDataParallel):
-            return model.module
+        if isinstance(model, DistributedDataParallel):
+            model = model.module
+        if isinstance(model, LightningDistributedModule):
+            model = model.module
```
Comment (carmocca) [inline pytorch_lightning/plugins/ddp_plugin.py:160]:
```suggestion
        if isinstance(model, (DistributedDataParallel, LightningDistributedModule)):
            model = model.module
```

**Review by carmocca — APPROVED** [review]

File: pytorch_lightning/plugins/ddp_plugin.py (hunk 11 above)
Comment (awaelchli) [reply pytorch_lightning/plugins/ddp_plugin.py:160]: @carmocca the way the if statements are layed out there sequentially is intentional, because when we use DDP the lightning module will be wrapped twice, so we have to do `model.module.module` in genreal :)

Comment (carmocca) [reply pytorch_lightning/plugins/ddp_plugin.py:160]: Touché

**Review by rohitgr7 — APPROVED** [review]

File: pytorch_lightning/overrides/data_parallel.py [hunk 12]
```diff
@@ -150,73 +152,75 @@ def parallel_apply(self, replicas, inputs, kwargs):
...
+
+        Args:
+            pl_module: the model to wrap
+
+        """
+        super().__init__()
+        self.module = pl_module
+
+    def forward(self, *inputs, **kwargs):
+        if self.module.training:
```
Comment (tchaton) [inline pytorch_lightning/overrides/data_parallel.py:190]: Any chance we could use the trainer state instead. If we user might change his model state by accident, we won't be calling the right function.

I am also thinking about people doing MC Dropout evaluation. `module.training` would be True if they don't set `training()`  properly.

Comment (justusschock) [reply pytorch_lightning/overrides/data_parallel.py:190]: I thought about that for quite a while, but I think, we should rely on the module attribute, but not from self.module but from self.

Users would probably only change it based on their LightningModule and when we rely on self.training and self.testing (probably has to be added here then) we should be fine in that regard and did not tie it to close to the trainer, since we try to get rid of all the trainer references everywhere.

File: pytorch_lightning/plugins/ddp_plugin.py (hunk 2 above)
Comment (justusschock) [reply pytorch_lightning/plugins/ddp_plugin.py:170]: We basically added this, since we did not want anything that is only DDP specific (i.e. any typechecks against the prior `LightningDistributedDataParallel` within the trainer/training-loop as this one should be backend agnostic.

**Review by justusschock — APPROVED** [review]

File: pytorch_lightning/overrides/data_parallel.py (hunk 12 above)
Comment (awaelchli) [reply pytorch_lightning/overrides/data_parallel.py:190]: I agree with both of you. The .training attribute is part of the nn.Module, however the .testing is set by the Trainer. This may not be so obvious, because there is no trace of training or testing attributes anywhere in the LightningModule. 
One attribute being part of nn.Module and one being part of LightningModule is strange and will not be so easy to debug. 

I'd like to keep it a strict refactor about the DDP class in this PR and not change the attributes yet.
Shall I create an issue so we can follow up on this?

Both of your suggestions seem reasonable to me. 
Justus's idea would require additional logic in the training loop to set the attributes on this wrapper.
Thomas's idea would basically be replacing self.module.training with self.module.trainer.training, right? We could also thing about adding read-only properties to the LightningModule as we did for other attributes that need to reference trainer.

**Review by williamFalcon — APPROVED** [review]

File: tests/overrides/test_data_parallel.py [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+
+    pl_module.training = False
+    pl_module.testing = True
+    dist_module(batch, batch_idx)
+    pl_module.test_step.assert_called_with(batch, batch_idx)
+
+    pl_module.training = False
+    pl_module.testing = False
+    dist_module(batch, batch_idx)
+    pl_module.validation_step.assert_called_with(batch, batch_idx)
```
Comment (Borda) [inline tests/overrides/test_data_parallel.py:30]:
```suggestion
    pl_module.training = True
    dist_module(batch, batch_idx)
    pl_module.training_step.assert_called_with(batch, batch_idx)

    pl_module.testing = True
    dist_module(batch, batch_idx)
    pl_module.test_step.assert_called_with(batch, batch_idx)

    pl_module.testing = False
    dist_module(batch, batch_idx)
    pl_module.validation_step.assert_called_with(batch, batch_idx)
```
after #5419 it is enum under hood

File: tests/overrides/test_data_parallel.py [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+
+    with pytest.warns(UserWarning, match="Your test_step returned None"):
+        pl_module.training = False
+        pl_module.testing = True
+        dist_module()
+
+    with pytest.warns(UserWarning, match="Your validation_step returned None"):
+        pl_module.training = False
+        pl_module.testing = False
+        dist_module()
```
Comment (Borda) [inline tests/overrides/test_data_parallel.py:55]:
```suggestion
        pl_module.training = True
        dist_module()

    with pytest.warns(UserWarning, match="Your test_step returned None"):
        pl_module.testing = True
        dist_module()

    with pytest.warns(UserWarning, match="Your validation_step returned None"):
        pl_module.testing = False
        dist_module()
```

**Review by Borda — APPROVED** [review]

File: tests/overrides/test_data_parallel.py (hunk 14 above)
Comment (Borda) [reply tests/overrides/test_data_parallel.py:55]: reverted...

File: tests/overrides/test_data_parallel.py (hunk 13 above)
Comment (Borda) [reply tests/overrides/test_data_parallel.py:30]: reverted...

File: tests/overrides/test_data_parallel.py [hunk 15]
```diff
@@ -0,0 +1,55 @@
...
+def test_lightning_distributed_module_methods():
+    """ Test that the LightningDistributedModule redirects .forward() to the LightningModule methods. """
+    pl_module = MagicMock()
+    dist_module = LightningDistributedModule(pl_module)
+
+    batch = torch.rand(5)
+    batch_idx = 3
+
+    pl_module.training = True
+    pl_module.testing = False
```
Comment (Borda) [inline tests/overrides/test_data_parallel.py:18]: curious how the module has these attributes, isn't it Trainer?

**Review by Borda — CHANGES_REQUESTED** [review]

File: tests/overrides/test_data_parallel.py (hunk 15 above)
Comment (awaelchli) [reply tests/overrides/test_data_parallel.py:18]: see my answer [here](https://github.com/PyTorchLightning/pytorch-lightning/pull/5185#discussion_r555001405)
The property is copied to the model [here](https://github.com/PyTorchLightning/pytorch-lightning/blob/404af43cde6696d04bb1899da2bb7e334e49716d/pytorch_lightning/trainer/connectors/model_connector.py)
happy to follow up on this in a next step. The test that I wrote there is just to make sure it works the same as before refactor.

File: pytorch_lightning/overrides/data_parallel.py (hunk 12 above)
Comment (Borda) [reply pytorch_lightning/overrides/data_parallel.py:190]: good, I would be just careful about adding too many new attributes to LightningModule :]

**Review by Borda — APPROVED** [review]
